### PR TITLE
Add tests for migration between source modes in logsdb data stream

### PR DESCRIPTION
**Pull request by lkts**
This PR adds tests for source mode migration.

### CODE REVIEW

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+public class LogsDbSourceModeMigrationIT extends LogsIndexModeRestTestIT {
+    public static final String INDEX_TEMPLATE = """
```
Comment (lkts) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:40]: I'll see if i can use start trial API in a follow up. Haven't tried it here.

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        );
+        assertThat(initialSourceMode, equalTo("stored"));
+        var initialIndexMode = (String) getSetting(client, getDataStreamBackingIndex(client, "my-logs-ds-test", 0), "index.mode");
+        assertThat(initialIndexMode, equalTo("logsdb"));
+
+        var indexedWithStoredSource = new ArrayList<XContentBuilder>();
+        var indexedWithSyntheticSource = new ArrayList<XContentBuilder>();
+        for (int i = 0; i < 10; i++) {
+            indexedWithStoredSource.add(generateDoc());
+            indexedWithSyntheticSource.add(generateDoc());
```
Comment (kkrik-es) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:149]: Shouldn't we be adding the same doc in both?

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java (hunk 1 above)
Comment (kkrik-es) [reply x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:38]: Right, these tests are a good start, but we need to test setting logsdb mode with various licenses and getting the expected source modes. We also need to check upgrading/downgrading, with conflicting settings in the index template (e.g. tempate says "synthetic" or null or nothing, license says "stored"). There are many combinations here, so we'd need to organize the testing logic well to keep the tests succinct.

**Review by kkrik-es — APPROVED** [review]

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        assertThat(finalSourceMode, equalTo("synthetic"));
+
+        Response syntheticSourceBulkResponse = bulkIndex(client, "my-logs-ds-test", indexedWithSyntheticSource, 10);
+        assertOK(syntheticSourceBulkResponse);
+        assertThat(entityAsMap(syntheticSourceBulkResponse).get("errors"), Matchers.equalTo(false));
+
+        var allDocs = Stream.concat(indexedWithStoredSource.stream(), indexedWithSyntheticSource.stream()).toList();
+
+        var sourceList = search(new SearchSourceBuilder().query(QueryBuilders.matchAllQuery()).size(allDocs.size()), "my-logs-ds-test");
+        assertThat(sourceList.size(), equalTo(allDocs.size()));
```
Comment (salvatore-campagna) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:173]: duplicate

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            indexedWithStoredSource.add(generateDoc());
+            indexedWithSyntheticSource.add(generateDoc());
+        }
+
+        Response storedSourceBulkResponse = bulkIndex(client, "my-logs-ds-test", indexedWithStoredSource, 0);
+        assertOK(storedSourceBulkResponse);
+        assertThat(entityAsMap(storedSourceBulkResponse).get("errors"), Matchers.equalTo(false));
+
+        assertOK(putComponentTemplate(client, "my-logs-migrated-source", SYNTHETIC_SOURCE_COMPONENT_TEMPLATE));
+        assertOK(rolloverDataStream(client, "my-logs-ds-test"));
```
Comment (salvatore-campagna) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:157]: When rolling over can we also check that the actual rollover happened by checking `rolled_over`?

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            "settings": {
+              "index": {
+                "mapping.source.mode": "synthetic"
+              }
+            }
+          }
+        }""";
+
+    @ClassRule()
+    public static ElasticsearchCluster cluster = ElasticsearchCluster.local()
```
Comment (salvatore-campagna) [inline x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:103]: If not using standard template we can disable them and avoid potential template composition issues by setting `stack.templates.enabled` to `false`.

File: x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java (hunk 2 above)
Comment (lkts) [reply x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java:149]: We will see the union of these two lists in search so we want them to be different to assert nothing gets lost.

**Comment by lkts** [comment]
@elasticmachine update branch

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/115374)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**